src/test/java/org/broadinstitute/hellbender/tools/recalibration/ContextCovariateUnitTest.java

```diff
@@ -29,31 +33,47 @@ public void initCache() {
+            SAMRecord clippedRead = ReadClipper.clipLowQualEnds(read, RAC.LOW_QUAL_TAIL, ClippingRepresentation.WRITE_NS);
+            ReadCovariates readCovariates = new ReadCovariates(read.getReadLength(), 1);
+            covariate.recordValues(read, readCovariates);
 
-    public static void verifyCovariateArray(int[][] values, int contextSize, SAMRecord read, Covariate contextCovariate) {
-        for (int i = 0; i < values.length; i++)
-            Assert.assertEquals(contextCovariate.formatKey(values[i][0]), expectedContext(read, i, contextSize));
+            verifyCovariateArray(readCovariates.getMismatchesKeySet(), RAC.MISMATCHES_CONTEXT_SIZE, clippedRead, covariate, RAC.LOW_QUAL_TAIL);
+            verifyCovariateArray(readCovariates.getInsertionsKeySet(), RAC.INDELS_CONTEXT_SIZE, clippedRead, covariate, RAC.LOW_QUAL_TAIL);
+            verifyCovariateArray(readCovariates.getDeletionsKeySet(), RAC.INDELS_CONTEXT_SIZE, clippedRead, covariate, RAC.LOW_QUAL_TAIL);
+        }
+    }
 
+    public static void verifyCovariateArray(int[][] values, int contextSize, SAMRecord read, Covariate contextCovariate, final byte lowQualTail) {
+        for (int i = 0; i < values.length; i++) {
+            Assert.assertEquals(contextCovariate.formatKey(values[i][0]), expectedContext(read, i, contextSize, lowQualTail), "offset " + i);
+        }
     }
 
-    public static String expectedContext (SAMRecord read, int offset, int contextSize) {
-        final String bases = stringFrom(read.getReadBases());
-        String expectedContext = null;
-        if (offset - contextSize + 1 >= 0) {
-            String context = bases.substring(offset - contextSize + 1, offset + 1);
-            if (!context.contains("N"))
-                expectedContext = context;
+    public static String expectedContext (final SAMRecord originalRead, final int offset, final int contextSize, final byte lowQualTail) {
+        final SAMRecord clippedRead = ReadClipper.clipLowQualEnds(originalRead, lowQualTail, ClippingRepresentation.WRITE_NS);
+        final boolean negStrand = clippedRead.getReadNegativeStrandFlag();
+
+        final byte[] strandedBaseArray = negStrand ? BaseUtils.simpleReverseComplement(clippedRead.getReadBases()): clippedRead.getReadBases();
+        final String strandedBases = stringFrom(strandedBaseArray);
+
+        final int strandedOffset = negStrand ? clippedRead.getReadLength() - offset - 1 : offset;
+        final int offsetOfContextStart = strandedOffset - contextSize + 1;
+        if (offsetOfContextStart < 0) {
+            return null;
+        } else {
+            final int offsetOneAfterContextEnd = offsetOfContextStart + contextSize;
+            final String context = strandedBases.substring(offsetOfContextStart, offsetOneAfterContextEnd);
+            if (context.contains("N")) {
+                return null;
+            } else {
+                return context;
+            }
```

Comment:
Is it fair to say that you've basically duplicated the code from the covariate classes in your tests to compute expected values? This is usually a sign of a poorly-designed test....you want to make assertions about the result without duplicating/re-implementing the thing you're testing.